### PR TITLE
rgw: lc support Content-MD5 request header and fix a rgw crash bug

### PR DESCRIPTION
**Pull request by ZVampirEM77**
1、According to AWS S3, this header must be used as a message integrity check to verify that the request body was not corrupted in transit. Content-MD5 is the base64-encoded 128-bit MD5 digest of the data

2、When a putting lc config request does not include an ID tag in lc configuration xml, RGW should generate a random ID for the lc configuration.
At present RGW will crash when generate a random ID for lc configuration.

Fixes: http://tracker.ceph.com/issues/21980

### CODE REVIEW

**Comment by ZVampirEM77** [comment]
@mikulely Thank you~ I have fix the loadgen issue in the PR https://github.com/ceph/ceph/pull/18536

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -4843,6 +4843,16 @@ void RGWPutLC::execute()
   RGWLCXMLParser_S3 parser(s->cct);
   RGWLifecycleConfiguration_S3 new_config(s->cct);
 
+  content_md5 = s->info.env->get("HTTP_CONTENT_MD5");
+  if (content_md5 == nullptr) {
```
Comment (theanalyst) [inline src/rgw/rgw_op.cc]: is AWS s3 behaviour similar, do they reject requests without a HTTP_CONTENT_MD5 for lifecycle...  not sure whether we should error out if no md5 is supplied

Comment (ZVampirEM77) [reply src/rgw/rgw_op.cc]: @theanalyst Yeah, if not supplying the `Content-MD5` in request headers, the response message of AWS S3 is
`<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InvalidRequest</Code><Message>Missing required header for this request: Content-MD5</Message>
...
</Error>`

File: src/rgw/rgw_lc.h [hunk 2]
```diff
@@ -152,6 +152,7 @@ WRITE_CLASS_ENCODER(LCFilter);
 class LCRule
 {
 protected:
+  CephContext *cct;
```
[the remark this answers is not in the document]
Comment (dang) [reply src/rgw/rgw_lc.h]: I agree with this.

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -4843,6 +4843,16 @@ void RGWPutLC::execute()
...
+    op_ret = -ERR_INVALID_REQUEST;
+    s->err.message = "Missing required header for this request: Content-MD5";
+    ldout(s->cct, 5) << s->err.message << dendl;
+    return;
+  }
+
+  std::string content_md5_bin = rgw::from_base64(boost::string_view(content_md5));
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: need to catch exceptions from `from_base64()`, which will throw if the input isn't base64

Comment (ZVampirEM77) [reply src/rgw/rgw_op.cc]: @cbodley Done

**Comment by ZVampirEM77** [comment]
@theanalyst @dang @cbodley @adamemerson I have updated my commits, please review this PR again. Thank you~

**Comment by ZVampirEM77** [comment]
@cbodley Done~

**Comment by ZVampirEM77** [comment]
Jenkins, retest this please.

**Comment by yuriw** [comment]
[wip-yuri-testing-2017-10-31-1636](https://trello.com/c/RZ136ACY/191-wip-yuri-testing-2017-10-31-1636)

**Pull request merged**